Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp

```diff
@@ -188,25 +192,121 @@ class ConstExprGenerator {
     PartialResult WARN_UNUSED_RETURN atomicFence(ExtAtomicOpType, uint8_t) CONST_EXPR_STUB
     PartialResult WARN_UNUSED_RETURN truncTrapping(OpType, ExpressionType, ExpressionType&, Type, Type) CONST_EXPR_STUB
     PartialResult WARN_UNUSED_RETURN truncSaturated(Ext1OpType, ExpressionType, ExpressionType&, Type, Type) CONST_EXPR_STUB
-    PartialResult WARN_UNUSED_RETURN addI31New(ExpressionType, ExpressionType&) CONST_EXPR_STUB
+
+    PartialResult WARN_UNUSED_RETURN addI31New(ExpressionType value, ExpressionType& result)
+    {
+        result = JSValue::encode(JSValue(static_cast<int>(value & 0x7fffffff)));
+        return { };
+    }
+
     PartialResult WARN_UNUSED_RETURN addI31GetS(ExpressionType, ExpressionType&) CONST_EXPR_STUB
     PartialResult WARN_UNUSED_RETURN addI31GetU(ExpressionType, ExpressionType&) CONST_EXPR_STUB
-    PartialResult WARN_UNUSED_RETURN addArrayNew(uint32_t, ExpressionType, ExpressionType, ExpressionType&) CONST_EXPR_STUB
-    PartialResult WARN_UNUSED_RETURN addArrayNewDefault(uint32_t, ExpressionType, ExpressionType&) CONST_EXPR_STUB
+
+    uint64_t createNewArray(uint32_t typeIndex, uint32_t size, uint64_t value)
+    {
+        return arrayNew(m_instance.get(), typeIndex, size, value);
+    }
+
+    PartialResult WARN_UNUSED_RETURN addArrayNew(uint32_t typeIndex, ExpressionType size, ExpressionType value, ExpressionType& result)
+    {
+        WASM_PARSER_FAIL_IF(!Options::useWebAssemblyGC(), "Wasm GC is not enabled");
+
+        if (m_mode == Mode::Evaluate)
+            result = createNewArray(typeIndex, static_cast<uint32_t>(size), value);
```

Comment:
Are we actually guaranteed that these will stay alive? When they are on the stack or in a register, they are fine, but what if they get stored only to a ResultList (which is a Vector<>)?

Reply:
Hi @justinmichaud is the situation that you are imagining something like the following?

```
(module
  (type (struct))
  (global (ref 0) (struct.new 0) (struct.new 0) (struct.new 0)) ;; doesn't validate, but suppose it does
)
```

In the global, there are three struct instances on the stack in the `ResultList`. This example doesn't validate, but let's say it did and only the last item on the stack was taken as the result. Then the other two structs are not reachable and get deallocated.

I think in practice, only init expressions where allocated values will be kept live pass validation. In particular, init expressions don't have `drop` so any allocated values must either be consumed (to be put in another object's fields) or returned from the init expr (and thus have a reference to them). And init exprs can't do any mutation of state, so you can't launder a reference out either.

Might be misunderstanding the concern though, happy to look into this more.